soc/arm/st_stm32/stm32g0/power.c

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2021 STMicroelectronics.
+ * Copyright (c) 2021 G-Technologies Sdn. Bhd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr.h>
+#include <pm/pm.h>
+#include <soc.h>
+#include <init.h>
+
+#include <stm32g0xx_ll_utils.h>
+#include <stm32g0xx_ll_bus.h>
+#include <stm32g0xx_ll_cortex.h>
+#include <stm32g0xx_ll_pwr.h>
+#include <stm32g0xx_ll_rcc.h>
+#include <stm32g0xx_ll_system.h>
+#include <clock_control/clock_stm32_ll_common.h>
+
+#include <logging/log.h>
+LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
+
+/* Invoke Low Power/System Off specific Tasks */
+void pm_power_state_set(struct pm_state_info info)
+{
+	if (info.state != PM_STATE_SUSPEND_TO_IDLE) {
+		LOG_DBG("Unsupported power state %u", info.state);
+		return;
+	}
+
+	switch (info.substate_id) {
+	case 1: /* this corresponds to the STOP0 mode: */
+		/* enter STOP0 mode */
+		LL_PWR_SetPowerMode(LL_PWR_MODE_STOP0);
+		LL_LPM_EnableDeepSleep();
+		/* enter SLEEP mode : WFE or WFI */
+		k_cpu_idle();
+		break;
+	case 2: /* this corresponds to the STOP1 mode: */
+		/* enter STOP1 mode */
+		LL_PWR_SetPowerMode(LL_PWR_MODE_STOP1);
+		LL_LPM_EnableDeepSleep();
+		/* enter SLEEP mode : WFE or WFI */
+		k_cpu_idle();
+		break;
+	default:
+		LOG_DBG("Unsupported power state substate-id %u",
+			info.substate_id);
+		break;
+	}
+}
+
+/* Handle SOC specific activity after Low Power Mode Exit */
+void pm_power_state_exit_post_ops(struct pm_state_info info)
+{
+	if (info.state != PM_STATE_SUSPEND_TO_IDLE) {
+		LOG_DBG("Unsupported power substate-id %u", info.state);
+	} else {
+		switch (info.substate_id) {
+		case 1:	/* STOP0 */
+			/* Clear SLEEPDEEP bit */
+			LL_LPM_EnableSleep();
+			break;
+		case 2:	/* STOP1 */
+			/* Clear SLEEPDEEP bit */
+			LL_LPM_EnableSleep();
+			/* 'Restores' the registers set when entering STOP1 */
+			LL_PWR_SetPowerMode(LL_PWR_MODE_STOP0);
```

Comment:
for stop1 : choose the LL_PWR_MODE_STOP**1**

Reply:
Here is where it exits STOP1 and "restores" the registers set by STOP1 when it first entered.

Reply:
Can you explain why this is required ?
In other series, we don't restore `LL_PWR_MODE_STOP0` when exiting `LL_PWR_MODE_STOP1`

Reply:
> Can you explain why this is required ?

I actually haven't try what would happen if I only restore the `SLEEPDEEP` and don't restore restore `LL_PWR_MODE_STOP0` when exiting `LL_PWR_MODE_STOP1`.

I'll try it tomorrow.

Reply:
I confirm that this 2 lines are not required
```c
/* 'Restores' the registers set when entering STOP1 */
LL_PWR_SetPowerMode(LL_PWR_MODE_STOP0);
```